#456921 [C#テキスト]8章演習問題

kurosu-being opened the pull request:
第8章　日付、時刻の操作

DateTime構造体を中心に日付と時刻の操作を学習しました。

kurosu-being committed:
#459603 [C#テキスト]8章演習問題

Practice8-1-1を実装しました。

kurosu-being committed:
#459603 [C#テキスト]8章演習問題

Practice8-1-2を仮に実装しました。
次の指定曜日を求めるメソッドをコメントにしました。

kurosu-being committed:
#459603 [C#テキスト]8章演習問題

Practice8-1-2を実装しました。
Practice8-1-3のファイルを追加しました。

kurosu-being committed:
#459603 [C#テキスト]8章演習問題

Practice8-2-1の不要なusingを削除しました。
Practice8-3-1を実装しました。

yuki-tanaka-being commented on 章末問題8/Practice8-3-1/TimeWatch.cs:
問題の意図が読めていないですが、`TimeWatch` は `DateTime` を使って実装するお題かと思いました。
`Stopwatch` みたいなクラスを自分で実装してみなさいということかな？
余裕があれば、`DateTime` を使って実装するパターンも挑戦してはいかがでしょうか。

---

あとは、指示通りなら `TimeWatch` を new した時点でカウント開始かなと思います。
お題ではなくちゃんとしたストップウォッチを実装するなら、今回実装したように `Start` が別である方がよいと思います。

kurosu-being replied:
ご指摘ありがとうございます！
StopWatchにStopメソッドを再実装するなどおかしなところがあった為、テキスト本来の主旨であるDateTimeクラスを用いた方法で修正いたしました。
また、テキストの指示に沿うようにTimeWatchをnewした時点でカウントが開始されるように修正しました。

yuki-tanaka-being replied:
いいですね。
機能としては問題ありません。

より良い設計の観点でリファクタリングの提案です。
実行中という情報を DRY にできます。
※DRYについては、[IT用語辞典](https://e-words.jp/w/DRY%E5%8E%9F%E5%89%87.html) や [DRY原則](https://xn--97-273ae6a4irb6e2hsoiozc2g4b8082p.com/%E3%82%A8%E3%83%83%E3%82%BB%E3%82%A4/DRY%E5%8E%9F%E5%89%87/) を参照。
`FIsRunning` という変数が実行中という状態を表していますが、実際に実行中にするには開始時刻を保持する必要があります。
逆を言うと、開始時刻を持っていれば実行中であると言えます。
そうなると、実行中という状態が二つの変数で管理されていることになります。
この二つの変数は依存しており、必ずセットで扱わないと正しい状態を示さないことになってしまいます。
今は単純な実装なのでバグを埋め込む可能性は低いですが、コード量が多くなるにつれて複雑度が増していきます。

それを解消するために、状態の表現を1つのオブジェクトで行うようにします。
実現方法は `FStartTime` がデフォルト値なら未実行、それ以外なら実行中とするか、`FStartTime` を Null許容型にして null なら未実行、null 以外なら実行中とします。
デフォルト値の実装例は以下です。
```csharp
class TimeWatch {
    private DateTime FStartTime;
    private bool IsRunning => this.FStartTime != default;

    public TimeWatch() => this.Start();

    public void Start() {
        if (this.IsRunning) return;
        this.FStartTime = DateTime.Now;
    }

    public TimeSpan Stop() {
        try {
            return this.IsRunning ? (DateTime.Now - this.FStartTime) : TimeSpan.Zero;
        } finally {
            this.FStartTime = default;
        }
    }
}
```
実装例のポイントは2つあります。
* `IsRunning` というプロパティを用意することで、実行中を DRY にします。
評価したい場所で都度 `this.FStartTime != default` とも書けますが、それだと状態の定義が記載場所に散らばあり、`FIsRunning` で管理していたころより状態管理の複雑度が上がってしまいます。
* `finally` にデフォルト値に戻す処理を書くことで、`Stop` メソッドが実行されたら必ず実行中でなくなるようにします。
もしエラーが起きても、オブジェクトの内部状態は可能な限り正しく保てるようにします。

kurosu-being replied:
ご指摘ありがとうございます！
ご提案いただいた方法で修正し、動作を確認しました。
（defaultの表記だけ、C#7.1以上の言語バージョンが必要とのことだったのでdefault→default(DateTime)と修正ました。現在使用しているC#言語バージョンは6.0のようです）
重複していたIsRunningとFStartTimeの役割を、依存関係を利用して重複を回避することが出来るのですね。
勉強になります✍

kurosu-being committed:
#459603 [C#テキスト]8章演習問題

仮実装です。

kurosu-being committed:
#459603 [C#テキスト]8章演習問題

Practice8-3-1においてTimeWatch_DateTimeクラスを追加しました。

kurosu-being committed:
#459603 [C#テキスト]8章演習問題

Practice8-3-1において、TimeWatch.cs,TimeWatch-DateTime.csのコメントを追加しました。
Practice8-3-1において、TimeWatch.csのンストラクタを修正しました。

kurosu-being committed:
#459603 [C#テキスト]8章演習問題

Practice8-3-1において、TimeWatch-DateTime.csを削除しました
Practice8-3-1において、TimeWatch.csをDateTimeを使って書き直しました。

kurosu-being committed:
#459603 [C#テキスト]8章演習問題

Practice8-3-1において、DateTimeクラスのコメントを追加しました。
不要なファイルSSSを削除しました。

kurosu-being committed:
#459603 [C#テキスト]8章演習問題

Practice8-3-1において、DateTimeクラスのコメントを修正しました。

kurosu-being committed:
#459603 [C#テキスト]8章演習問題

Practice8-3-1において、DateTimeクラスの状態表現を1つのオブジェクトで表現できるように修正しました。

k-miyahara-being commented on 章末問題8/Practice8-1-1/Program8-1-1.cs:
```diff
@@ -0,0 +1,24 @@
+            var wJapanDay = wDate.ToString("ggyy年M月d日", wCulture);
+            var wDayOfWeek = wCulture.DateTimeFormat.GetDayName(wDate.DayOfWeek);
+
+            Console.WriteLine(wDate.ToString("f"));
```
1つめの表示形式の日付部分が満たせていないようです。
フォーマットを見直してみてください。

kurosu-being replied:
ご指摘ありがとうございます！
正しいフォーマットで表示されるよう修正しました。

k-miyahara-being commented on 章末問題8/Practice8-1-1/Program8-1-1.cs:
```diff
@@ -0,0 +1,24 @@
+            var wCulture = new CultureInfo("ja-JP");
+            wCulture.DateTimeFormat.Calendar = new JapaneseCalendar();
+            var wJapanDay = wDate.ToString("ggyy年M月d日", wCulture);
+            var wDayOfWeek = wCulture.DateTimeFormat.GetDayName(wDate.DayOfWeek);
```
ローカル変数の宣言は、使用する場所になるべく近いところで行うのがよいです。
メソッドの先頭で宣言された変数が、いろいろと処理を挟んだ後に末尾で使用されていると、
上から読んだときに余計な思考を生んでしまいます。

kurosu-being replied:
ご指摘ありがとうございます！
変数の宣言位置を変更し、使用する場所と対応するよう修正しました。
出来るだけ狭い範囲でコードを確認できるようにする為にも、宣言の位置に気をつけたいと思います。

k-miyahara-being commented on 章末問題8/Practice8-2-1/Program8-2-1.cs:
```diff
@@ -0,0 +1,24 @@
+        /// <param name="vDayOfWeek">指定曜日</param>
+        /// <returns></returns>
+        public static DateTime NextWeek(DateTime vDate, DayOfWeek vDayOfWeek) {
+            int wDaysSpan = ((int)vDayOfWeek - (int)vDate.DayOfWeek + 7) % 7;
+            return vDate.AddDays(wDaysSpan + 7);
```
計算式を見直してみてください。
今日が水曜日として、指定曜日を月または火曜日にすると、再来週の日にちになってしまいます。

kurosu-being replied:
ご指摘ありがとうございます！
計算式を修正し、来週の曜日が出るように修正しました。
また、コード上で曜日を選択する方法をやめ、ユーザーが曜日を指定するように変更いたしましたので、
ご確認いただけると幸いです。

k-miyahara-being replied:
こうしたらいいかも、を実際にやってみるのはとてもいいですね！
ぜひ今後も、学習のためにいろいろな実装を試してみてください👍

kurosu-being committed:
#459603 [C#テキスト]8章演習問題

Practice8-1-1において、コンソール1行目の表記が正しく表記されるよう修正しました。
Practice8-1-1において、変数の宣言位置を修正しました。
Practice8-2-1において、式を修正しました。

kurosu-being committed:
#459603 [C#テキスト]8章演習問題

Practice8-2-1において、ユーザーが数字で曜日を指定できるように修正しました。

kurosu-being committed:
#459603 [C#テキスト]8章演習問題

Practice8-2-1において、コンソールに表示されるメッセージを修正しました。

kurosu-being committed:
#459603 [C#テキスト]8章演習問題

Practice8-2-1において、フォーマットの修正を行いました。

kurosu-being committed:
#459603 [C#テキスト]8章演習問題

Practice8-2-1において、条件分岐の表記を変更しました。

kurosu-being committed:
#459603 [C#テキスト]8章演習問題

Practice8-2-1において、条件分岐の表記を再変更しました。
